tempesta_fw/http_parser.c

```diff
@@ -1325,28 +1331,64 @@ tfw_http_parse_req(void *req_data, unsigned char *data, size_t len)
 	 *
 	 * We must not rewrite abs_path, but still can cast host part
 	 * to lower case.
+	 *
+	 * RFC 3986 chapter 3.2:
+	 * authority = [userinfo@]host[:port]
+	 *
+	 * We parse authority. It can be "host" or "userinfo@host"
+	 * (port parsed later). At the begining we don't know,
+	 * which of variants we have.
+	 *
+	 * So we start to fill both req->host and req->userinfo. If we find
+	 * '@' we reset req->host and finish req->userinfo. If we don't,
+	 * we clean req->userinfo.
 	 */
-	__FSM_STATE(Req_UriHostStart) {
+	__FSM_STATE(Req_UriAuthorityStart) {
 		if (likely(isalnum(c) || c == '.' || c == '-')) {
 			*p = LC(*p);
+			tfw_http_msg_set_data(msg, &req->userinfo, p);
 			tfw_http_msg_set_data(msg, &req->host, p);
-			__FSM_MOVE_f(Req_UriHost, &req->host);
+			__FSM_MOVE_f(Req_UriAuthority, &req->host);
 		}
 		return TFW_BLOCK;
 	}
 
-	__FSM_STATE(Req_UriHost) {
-		if (likely(isalnum(c) || c == '.' || c == '-')) {
+	__FSM_STATE(Req_UriAuthority) {
+		if (likely(isalnum(c) || c == '.' || c == '-' || c == '@')) {
 			*p = LC(*p);
-			__FSM_MOVE_f(Req_UriHost, &req->host);
+
+			if (unlikely(c == '@')) {
+				if (!TFW_STR_EMPTY(&req->userinfo)) {
+					TFW_DBG3("Second '@' in authority\n");
```

Comment:
This message is about reason to block the request, so it should be more important. Please use TFW_DBG